### ISSUE TITLE
Brackets turns into a web browser

### ISSUE THREAD

**Issue by dangowans** [adobe/brackets#12441]
While browsing the web in Firefox 46.0.1, I accidentally clicked and dragged a link from a webpage into Brackets.

When I let go of the link in Brackets, the editor disappeared without warning, and the window became a web browser.

Dragging this [Google Link](http://www.google.com) into the Brackets window has the same behaviour.

There's no consistent way to get back to the editor without closing Brackets completely and reopening it.  Any unsaved changes before the click-and-drag event are lost.  Menu options stop working too.

![Google Inside of Brackets](https://cloud.githubusercontent.com/assets/19495149/15437111/4ea7b880-1e93-11e6-940f-cc42f52ca76e.png)

I'm running Brackets 1.6 on Windows 7.  

(The same thing happens from Google Chrome 50.)

**Comment by petetnt** [comment]
Hi @dangowans,

this issue is a duplicate of #11970 (and some other older ones). Due to the nature of the component that backs Brackets (CEF) this thing is bound to happen, but the fix should be relatively simple. If you feel like putting in a PR I can guide you through the steps (or I'll try to get around to it later myself).

**Comment by dangowans** [comment]
Honestly, I only created an account here to submit this bug.  I would have to do a whole lot of reading up on git before I'd totally understand what a PR is, let alone be comfortable "putting in a PR".

**Comment by petetnt** [comment]
@dangowans That's alright! Thanks for reporting the issue! 👍

**Comment by wasifhyder** [comment]
Hey. I'd like to work on this. Can I?

**Comment by petetnt** [comment]
Hi @wasifhyder,

of course! If this is your first time contributing to Brackets, check out the guide on [How to hack on Brackets first](https://github.com/adobe/brackets/wiki/How-to-Hack-on-Brackets) first, which outlines all the the steps you need from getting started to putting up an PR. 

If you have any questions (about the process or the issue), just post them here! Have fun 👍

**Comment by jamran7** [comment]
Hi,
I am working on brackets in a separate repository with a team as part of a college course project. I've already started a topic about the solution with brackets-dev: "Brackets turning into browser issue https://github.com/adobe/brackets/issues/12441 (may be old)". Is there any drag-and-drop copy for text in Brackets? Brackets currently stops the browser's default behavior only when there is data to transfer like a file, but if you drag text into Brackets such as a url then Brackets will open that url unless the default behavior is halted. The only problem with my solution is that when drag-and-drop is handled, the default behavior of the browser is halted all together before checking if there is any data to transfer which would disable any drag-copy of text; that's the reason why I'm asking.

**Comment by petetnt** [comment]
Hi @jamram7,

Go to `Debug -> Open Preferences File` and add `"dragDropText": true` to the `brackets.json` (right panel) to enable drag and drop text functionality.
`

**Comment by jamran7** [comment]
Huh, even with that on and after I reverted the code back to the original state, I can't drag copy text to brackets. What is also weird is that with dragText set to true I can't drag a url into brackets and expect it to produce the bug either. However I can drag a url into the extension plane and produce the bug with dragText set to true which makes me wonder if my solution is really a solution.

**Comment by jamran7** [comment]
Ok so I tried to draging-and-droping url's all around the editor with my solution working with brackets and it does not allow the browser to takeover, even with the split views, but the issue with dragAndDrop not working when it is true bothers me. I'm working on a Mac and I've tried it with a clean install, not a git link, on my Windows desktop and it too doesn't drag-and-drop text. Can someone verify that they can drag-copy text with the dragAndDrop option set to true?

**Comment by redmunds** [comment]
The `dragAndDrop` option allows dragging selected text to _move_ it. To _copy_ selected text, hold down Cmd key on Mac (Ctrl key on Windows,Linux).

**Comment by jamran7** [comment]
So I've had to change the solution, but it's where it should be now; I've tested the drag-and-drop option and it works with text. You can drag-copy text just not urls, and it works in the different split views. The solution is this: 
```
            var types = event.dataTransfer.types;
            if(!files.length){// stop default behavior if a url is dragged in so the browser does not takeove
                types.forEach(function (value){
                    if(value == "text/uri-list"){ //plain text just has text/html
                        event.stopPropagation();
                        event.preventDefault();
                        return;
                    }
                });
            }
```
And you put that before `if (files && files.length) {` in two functions `function handleDragOver(event) {` and `function handleDrop(event) {`. These two functions are in the function `function attachHandlers() {` under utils/DragAndDrop.js.

**Comment by petetnt** [comment]
Good job @jamran7!

Would you be interested in making that a PR? See https://github.com/adobe/brackets/blob/master/CONTRIBUTING.md for more information.

**Comment by jamran7** [comment]
Yes I am, I am running the test now and will soon create a PR.

**Comment by ficristo** [comment]
@jamran7 from your solution it seems to me you are implementing the [allowDropFileTypes](https://codemirror.net/doc/manual.html#option_allowDropFileTypes) option of CodeMirror.
Could you investigate if it is the case?
Keep in mind I didn't look at this issue at all so maybe it is the wrong route, I still believe it's worth double checking.

**Comment by jamran7** [comment]
> allowDropFileTypes: array<string>
> When set (default is null) only files whose type is in the array can be dropped into the editor. The strings should be MIME types, and will be checked against the type of the File object as reported by the browser.

CM is apparently not in charge of handling files in Brackets which is why Brackets pre-empts CM. CM wouldn't see any files anyways if that option was used. (Scroll, the cut off threw me off)
```
// Over CodeMirror specifically, always pre-empt CodeMirror's drag event handling if files are being dragged - CM stops
        // propagation on any drag event it sees, even when it's not a text drag/drop. But allow CM to handle all non-file drag
        // events. See bug #10617.
        window.document.body.addEventListener("dragover", function (event) {
            if ($(event.target).closest(".CodeMirror").length) {
                handleDragOver(event);
            }
        }, true);
        window.document.body.addEventListener("drop", function (event) {
            if ($(event.target).closest(".CodeMirror").length) {
                handleDrop(event);
            }
        }, true);
```

**Comment by ficristo** [comment]
@jamran7 thank you. I really appreciate.

**Comment by jamran7** [comment]
The bug fix passes all tests, with the exception of one suite which seems to be a bug with the test, and it's only one test of the whole suite. It said it was expecting <!doctype html> in the test file, image is attached. As for JSLint the changed code follows with the exception that it want's "use strict" in every function.
 
<img width="894" alt="screen shot 2016-11-07 at 1 21 15 pm" src="https://cloud.githubusercontent.com/assets/21205059/20633486/155ad852-b315-11e6-97e5-70845a7ef1e1.png">

**Comment by ficristo** [comment]
The `Find in Files` error is sort of known, don't worry too much about it.
If your PR is ready, just submit it so we can take a look: without seeing the code we cannot give more advices about the JSLint warning.

**Comment by 4O4** [comment]
This has been fixed in PR #12946
